proposals/nnnn-screen-manager-play-audio.md

```diff
@@ -0,0 +1,119 @@
+
+   PlayAudioData(@NonNull SdlFile audioFile)
+   PlayAudioData(@NonNull String spokenString)
+   PlayAudioData(@NonNull String phoneticString, @NonNull SpeechCapabilities phoneticType)
+}
+```
+
+### Screen Manager API
+Next is the addition to the screen manager public API itself to play the audio data.
+
+##### iOS
+```objc
+@interface SDLScreenManager: NSObject
+// Everything already there
+
+/**
+Play some audio on the module.
+
+If the audio contains a file that needs to be uploaded, the file will be uploaded before it is played.
+
+The handler will be called when the module responds to the request to play.
+*/
+- (void)playAudio:(NSArray<SDLPlayAudioData *> *)audioData withCompletionHandler:(SDLScreenManagerUpdateCompletionHandler)handler;
+
+@end
+```
+
+##### Java
+```java
+public class BaseScreenManager {
+    // Everything already there
+    public void playAudio(List<PlayAudioData> audioData, CompletionListener listener)
+}
+```
+
+### JavaScript Suite APIs
+Due to the size of the iOS APIs and the similarity between the iOS, Java Suite and eventual JavaScript Suite APIs, this proposal does not present the public APIs of the JavaScript Suite APIs – especially because the JavaScript Suite APIs do not currently have a screen manager layer. The JavaScript Suite APIs should mirror the iOS and Java Suite API appropriately and is up to the maintainers' discretion. However, if any changes needed to be made such that they impacted the iOS / Java Suite API (such as the alteration, addition, or removal of a method or property), then a proposal revision would be needed.
+
+### General Notes
+* There are two general possibilities for how to handle if the developer calls `playAudio` while another piece of audio is playing.
+    1. We could queue the next `playAudio` and wait for the first to finish. We would use a queueing system similar to the `ChoiceSetManager` for this.
+    2. We could send the next `playAudio` `Speak` RPC immediately and allow the head unit to handle playing it after the first, or, more likely, error and then return the error.
+    * Choice #1 should be implemented in this proposal.
+
+## Potential downsides
+This adds some complexity by needing to handle file uploads for possible `TTSChunk` file uploads. However, this is complexity that must currently be handled by the developer.
+
+## Impact on existing code
+This will be a minor version change for all app libraries.
+
+## Alternatives considered
+1. We could move `playAudio` to the `SDLManager` instead of the `ScreenManager` because it doesn't _technically_ change the screen. However, the `ScreenManager` is already set up for sub-managers, which this will use. Furthermore, the developer usually looks on the `ScreenManager` for their primary app implementation.
+2. We could implement option #2 of point 1 instead of option #1, but the developer will need to track the responses of their `playAudio` calls manually to ensure there are no conflicts.
```

Comment:
This may need to be updated based on list formatting changes to the `General Notes` section above.

Reply:
Updated!